src/main/java/frc/robot/commands/autonomous/TrajectoryCommandGroup.java

```diff
@@ -0,0 +1,188 @@
+package frc.robot.commands.autonomous;
+
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.trajectory.Trajectory;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.CommandGroupBase;
+import edu.wpi.first.wpilibj2.command.RamseteCommand;
+import frc.robot.subsystems.DrivetrainSubsystem;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+/**
+ * A command group that executes commands while driving along a trajectory.
+ *
+ * @author Zane Othman-Gomez
+ */
+public class TrajectoryCommandGroup extends CommandGroupBase {
+  private final RamseteCommand m_trajectoryCommand;
+  private final DrivetrainSubsystem m_drive;
+
+  // A map of every command that is part of this group
+  private final Map<Command, PositionTrigger> m_allCommands = new HashMap<>();
+
+  // A map of commands that haven't executed
+  // This is kept to avoid triggering a command multiple times
+  private final Map<Command, PositionTrigger> m_untriggeredCommands = new HashMap<>();
+
+  // A list of currently executing commands (for ending purposes)
+  private final List<Command> m_executingCommands = new ArrayList<>();
+
+  public TrajectoryCommandGroup(
+      Trajectory trajectory, DrivetrainSubsystem drive, Map<Command, PositionTrigger> commands) {
+    // Create a command to follow the provided trajectory and inherit its requirements
+    m_trajectoryCommand = new FollowTrajectory(drive, trajectory);
+    m_requirements.addAll(m_trajectoryCommand.getRequirements());
+
+    // The trajectory command shouldn't be able to be used elsewhere (although this is probably
+    // unnecessary since it's instantiated here)
+    requireUngrouped(m_trajectoryCommand);
+
+    // Store a reference to the drivetrain subsystem for fetching the robot's pose
+    m_drive = drive;
+
+    // Add the commands to this command group
+    addCommands(commands);
+  }
+
+  @Override
+  public void initialize() {
+    // Clear and repopulate the unfinished command map
+    m_untriggeredCommands.clear();
+    m_untriggeredCommands.putAll(m_allCommands);
+
+    // Clear the executing command list
+    m_executingCommands.clear();
+
+    // Initialize the trajectory following command
+    m_trajectoryCommand.initialize();
+  }
+
+  @Override
+  public void execute() {
+    // Drive along the trajectory
+    m_trajectoryCommand.execute();
+
+    // A ListIterator allows for the modification of a list while iterating over it
```

Comment:
When is this occurring again? (modifying while iterating)

Reply:
When a command finishes it's removed from the currently executing list (this actually happens on [line 86](https://github.com/AEMBOT/FRC_2022/pull/6/files/dfc8283251690b391a0969e2b1720484ed818ac1#diff-78c97e56a7335e440b23b0c7f24702eae436bce46ba902af90caf1a7eb98f1f9R86) of TrajectoryCommandGroup.java). I could also store the finished commands in a separate list and remove them after finishing iterating, but removing them during the iteration made more sense in my head.